Add password support for ldap settings in admin-default.json

luhahn opened the pull request:
This PR also introduces the ability to upgrade the database with the admin-default.json

Once a admin-default.json is present it will override the settings. This is needed in order to support an upgradeable helm chart.

However i know that this might not be the optimal solution. I also thought about using the admin-default.json  as it is now and support an admin.json file in the root directory which will support the upgrade/override feature described above.

This PR referes to Issue: #729

luhahn committed:
add password as supported parameter for ldap

luhahn committed:
Change load order in favor of admin-default.json

luhahn commented:
So I just finished the helm chart and it works fine with this patch.

https://github.com/novumrgi/helm/tree/master/charts/glowroot

luhahn commented:
@trask Since there was no feedback, is a helm chart interesting at all?

trask commented:
hey @luhahn! sorry for delay, i will try to catch up on the open PRs in the next week

luhahn commented:
@trask Hope you're okay? Are there any news when this can be merged?

trask commented on central/src/main/java/org/glowroot/central/repo/CentralRepoModule.java:
```diff
@@ -93,19 +91,18 @@ public CentralRepoModule(ClusterManager clusterManager, Session session, File co
-        }
+        File adminDefaultFile = new File(confDir, "admin-default.json");
 
+        if (adminDefaultFile.exists()) {
```
this change will cause the central configuration to be overwritten on every startup

luhahn replied:
Ok, this somehow went totally out of my scope. Sorry for the major delay here.

This would be needed in order for the helm chart to work and to be able to configure glowroot again when values have changed.
With using the helm chart we can easily deploy glowroot into k8s

luhahn replied:
Would there be a possibility to somehow allow this? Maybe with an additional starting flag or something like this?

trask replied:
ya, we'd need some way to know that the user wants to overwrite their config with `admin-default.json`.

what about just updating `admin.json` directly from the helm chart, instead of updating `admin-default.json`?

luhahn replied:
that could work as well, I'll give it a shot and report back here :) Thanks for the suggestion!

luhahn replied:
Ok, had some time to spare today and i tested a bit. However, this might be a stupid question, where exactly would the admin.json be located? I can't find it.

trask replied:
oh! my bad, I got the standalone vs the central collector confused. `admin.json` is only for standalone. central collector stores the admin configuration in cassandra. 🤔

luhahn replied:
Good, I'm not totally confused :D.
So the only way to configure a glowroot running cassandra is by using the admin-default. 
How about an external flag, which signals the user, that his config will always be overwritten by the admin-default. Something like a "helm-chart mode" :)

luhahn replied:
@trask which way would you prefer?

trask replied:
> How about an external flag, which signals the user, that his config will always be overwritten by the admin-default

this makes sense to me. in this mode `admin-default.json` would be the source of truth, and configuration will be overwritten with `admin-default.json` on each startup.

trask commented:
sure, both things sound good to me

* get the changes needed into this repo so @luhahn doesn't need to maintain/build a fork anymore
* move the helm chart itself to this repo and add @luhahn to [CODEOWNERS](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) for it (I think this fits this repo's scope since various other deployment artifacts are also published from this repo, e.g. docker and yum)

luhahn commented:
> sure, both things sound good to me
> 
>     * get the changes needed into this repo so @luhahn doesn't need to maintain/build a fork anymore
> 
>     * move the helm chart itself to this repo and add @luhahn to [CODEOWNERS](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) for it (I think this fits this repo's scope since various other deployment artifacts are also published from this repo, e.g. docker and yum)

we could include it here, though we might need another repo, where we deploy the artifact. The easiest way to do this is by creating a github page including an index.yaml, which references the actual artifact. Users can then use, for example [novumrgi repo](https://github.com/novumrgi/helm), the following commands:

```
helm repo add novum-rgi https://novumrgi.github.io/helm/
helm install my-release novum-rgi/<chart>
```

robinroos commented:
I think it would look better for there to be a /helm repo within the glowroot project, e.g.

helm repo add glowroot https://glowroot.github.io/helm/
helm install 14.0.0 glorwoot/glowroot-central

luhahn committed:
Add flag for helm mode

luhahn commented:
@trask alright, so I added a new property called helmMode into glowroot-central.properties. Since I'am not a java dev, I'am not sure if this is a fitting approach, happy for input :)

luhahn commented:
@trask I think you're currently as busy as me, but can you give me a short feedback, if the new commit would be an acceptable solution?

trask reviewed (commented):
yes, this looks good

robinroos commented:
@luhahn Hi there, I know it's been a while but I am still reliant on the novum-rgi helm chart and was installing glowroot from your "openshift-fix" tag just yesterday.  Can we bring this PR to fruition?  Thanks so much, Robin.

luhahn commented:
@robinroos sorry, I switched departments in the company and currently there is noone actively working on our helm charts glowroot. However yesterday i had time and updated my changes to the 0.14.2 release.

Unfortunatly i've got very little time to spare, but lets see if we can merge this :)

luhahn commented on common2/src/main/java/org/glowroot/common2/config/LdapConfig.java:
```diff
@@ -50,6 +50,12 @@ public String username() {
         return "";
     }
 
+    @Value.Default
+    @JsonInclude(Include.NON_EMPTY)
+    public String password() {
+        return "";
+    }
+    
```
i think it did not work otherwise. Have to test it :)

luhahn committed:
Correct boolean parameter for helm mode

luhahn commented:
after a bit of testing i noticed that with 0.14.2 there is the datastax driver and configfile introduced, which basically breaks the entire cassandra configuration i had in my helm chart previously.

i configured it via environment variables CASSANDRA_USERNAME, CASSANDRA_PASSWORD, etc. it seems this is no longer possible, so i have to see how to adjust it accordingly

Never-lab commented:
Triage ping — this still looks **relevant** (LDAP password in `admin-default.json` / Helm-friendly admin defaults; related #729).

Before maintainers re-review, could you please:
1. Rebase onto current `main` and confirm CI is green
2. Re-check that `admin-default.json` override / helm-mode behavior still matches today’s Central config loading
3. Confirm whether the password field encryption path is still correct vs current `LdapConfig`

@luhahn thanks if you can refresh this. @nowheresly — candidate to keep once rebased/validated.
